src/main/java/com/google/firebase/database/connection/WebsocketConnection.java

```diff
@@ -242,94 +245,65 @@ public void run() {
     };
   }
 
-  private boolean isBuffering() {
-    return frameReader != null;
+  /**
+   * Closes the low-level connection, and notifies the higher layer ({@link Connection}.
+   */
+  private void closeAndNotify() {
+    close();
+    delegate.onDisconnect(everConnected);
```

Comment:
I wonder if there should be an isClosed check before calling the delegate, unless you're sure that closeAndNotify() can't be called more than once or that the delegate will be tolerant of multiple callbacks.

Reply:
Delegate is idempotent with this regard. Once the connection has been marked closed at that level, it will start ignoring `onDisconnect()` events. But I think having that check makes the logic easier to follow. Added.